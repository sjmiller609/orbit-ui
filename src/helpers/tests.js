import { BrowserRouter } from 'react-router-dom'
import { shape } from 'prop-types'
import { mount } from 'enzyme'

// Instantiate router context
const router = {
  history: new BrowserRouter().history,
  route: {
    location: {},
    match: {},
  },
}

const createContext = () => ({
  context: { router },
  childContextTypes: { router: shape({}) },
})

// wrap with router context - to mount RR links correctly
export function mountWrap(node) {
  return mount(node, createContext())
}
