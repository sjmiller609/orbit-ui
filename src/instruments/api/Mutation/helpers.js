const has = (str, find) => ~str.toLowerCase().indexOf(find.toLowerCase())

export const formErrors = ({ error, errors }) => {
  if (!error) return

  let err = JSON.stringify(error).toLowerCase()
  return errors.find(e => {
    if (e.key && has(err, e.key)) return e
    else if (e.allKeys && e.allKeys.every(k => has(err, k))) return e
    else if (e.someKeys && e.someKeys.some(k => has(err, k))) return e
  })
}
