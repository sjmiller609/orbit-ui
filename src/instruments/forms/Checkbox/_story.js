import React from 'react'
import { storiesOf } from '@storybook/react'
import { Checkbox } from 'instruments'

storiesOf('Instruments|Forms.Checkbox', module)
  .add('Default', () => (
    <Checkbox
      id="mockCheckbox"
      name="mockCheckbox"
      onChange={() => null}
      onBlur={() => null}
      updateErrors={() => null}
      setRef={() => null}
      label="Mock checkbox"
      value="mockCheckbox"
    />
  ))
  .add('Required', () => (
    <Checkbox
      id="mockCheckbox"
      name="mockCheckbox"
      onChange={() => null}
      onBlur={() => null}
      updateErrors={() => null}
      setRef={() => null}
      label="Mock checkbox"
      value="mockCheckbox"
      require={true}
    />
  ))
