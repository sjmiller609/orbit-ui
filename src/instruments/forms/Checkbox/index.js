'use strict'
import React from 'react'
import PropTypes from 'prop-types'
import s from './styles.scss'
import classNames from 'classnames'
import { Row } from '../../../instruments'

class Checkbox extends React.Component {
  change = this.change.bind(this)
  onChange = this.onChange.bind(this)
  timeout = null
  validate = this.validate.bind(this)
  showError = this.showError.bind(this)

  state = {
    id: this.props.id || this.props.name,
    showError: false,
    touched: null,
    checked: !!this.props.value,
  }

  componentDidMount() {
    const { focus, value } = this.props
    if (focus) this.field.focus()
    this.validate({ value }) // adds required fields to form
  }

  componentWillReceiveProps({ value, error }) {
    const set = {}

    if (value !== this.props.value) {
      set.showError = false
      // run validation
      this.validate({ value, error })

      // show error after 3 seconds of stopped typing
      if (this.timeout) clearTimeout(this.timeout)
      this.timeout = setTimeout(() => {
        this.showError()
        clearTimeout(this.timeout)
      }, 3000)

      set.touched = true

      if (!!value && !this.state.checked) set.checked = true
      else if (!value && this.state.checked) set.checked = false
    }

    this.setState(set)
  }
  // should probably switch this to pull the props that don't change, and then json.stringify
  // NOTE: if any new props are passed in that are meant to be reactive, won't work.
  shouldComponentUpdate({ value, error }, { showError, touched }) {
    if (value !== this.props.value) return true
    if (error !== this.props.error) return true
    if (showError !== this.state.showError) return true
    if (touched !== this.state.touched) return true
    return false
  }

  componentWillUnmount() {
    clearTimeout(this.timeout)
  }

  onChange(value) {
    const { name, onChange } = this.props
    onChange(name, value)
  }

  change(e) {
    if (e) e.preventDefault()
    const checked = !this.state.checked
    this.setState({ checked })
    this.onChange(checked)
  }
  showError(showError = true) {
    if (showError === this.state.showError) return
    this.setState({ showError })
  }

  validate({ value }) {
    const { validate, name, updateErrors, required } = this.props
    let e

    if (required && (typeof value !== 'boolean' && !value)) {
      e = name.charAt(0).toUpperCase() + name.slice(1) + ' is required'
    }
    if (value || typeof value === 'boolean') {
      if (validate) {
        e = validate(value)
      }
    }

    updateErrors(name, e)
  }

  render() {
    const { className, label, name, value, required } = this.props
    const { checked } = this.state

    return (
      <Row
        justify="flex-start"
        align="flex-start"
        auto
        className={classNames(
          s.checkbox,
          required ? s.required : null,
          className
        )}>
        <input
          type="checkbox"
          name={name}
          checked={checked}
          value={value || ''}
          className={s.input}
          required={required}
          onChange={this.onChange}
        />

        <div
          className={classNames(s.box, checked ? s.on : s.off)}
          onClick={this.change}>
          <div className={s.check} />
        </div>

        {label && <label>{label}</label>}
      </Row>
    )
  }
}

Checkbox.propTypes = {
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  name: PropTypes.string.isRequired,
  id: PropTypes.string,
  onChange: PropTypes.func,
  error: PropTypes.string,
  className: PropTypes.string,
  updateErrors: PropTypes.func,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool,
  ]),
  focus: PropTypes.bool,
  validate: PropTypes.func, // check if isValid, return error message
  required: PropTypes.bool,
}

export default Checkbox
