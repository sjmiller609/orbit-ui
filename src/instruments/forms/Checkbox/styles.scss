@import 'src/styles/mixins.scss';

.checkbox {
  font-size: 0.85em;
  line-height: 1.4em;
  margin: 1rem 0 1.5rem;

  .box {
    flex-shrink: 0;
    border-radius: 3px;
    border: 3px solid $beige;
    width: 1.4em;
    height: 1.4em;
    cursor: pointer;
    position: relative;
    background-color: rgba($beige, 0);
    transition: background-color 140ms ease-in;
    //	@include hoverScaleSm;

    .check {
      border-bottom-width: 3px;
      border-color: $darkWhite;
      border-left: 0;
      border-right-width: 3px;
      border-style: solid;
      border-top: 0;
      content: '';
      height: 1em;
      position: absolute;
      top: -0.08em;
      left: 0.25em;
      transform: rotate(0) scale(0);
      transform-origin: center;
      width: 0.5em;
      transition: transform 140ms ease-in;
    }

    &.on {
      background-color: $beige;

      &:hover .check {
        transform: rotate(45deg) scale(0.9);
      }

      .check {
        transform: rotate(45deg) scale(1);
      }
    }
  }

  .input {
    height: 0;
    opacity: 0;
    overflow: hidden;
    position: absolute;
    width: 0;
  }

  label {
    line-height: inherit;
    vertical-align: middle;
    margin-left: 0.7em;
    font-size: 0.75em;
    color: $beige;

    @include font($f500);

    a {
      color: $green;

      @include font($f700);

      font-size: 1em;
    }
  }
}
