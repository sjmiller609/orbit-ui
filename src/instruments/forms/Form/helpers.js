// faltten and stringify nested keys
export const unpack = (obj, ns, hash) => {
  if (!obj) return
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keypath = ns ? `${ns}.${key}` : key
    if (value && typeof value === 'object') {
      return unpack(value, keypath, acc)
    }
    acc[keypath] = value
    return acc
  }, hash || {})
}

const roll = (keys, obj, value) => {
  const key = keys[0]
  // case 1 - not obj
  if (keys.length === 1) {
    obj[key] = value
    return obj
  }
  // case 2 - doesn't exist
  if (!obj[key]) {
    obj[key] = isNaN(keys[1]) ? {} : []
  }
  // case 3 - exists
  obj[key] = roll(
    keys.slice(1),
    Array.isArray(obj[key]) ? [...obj[key]] : { ...obj[key] },
    value
  )

  return obj
}

// convert back into obj
export const pack = (obj, type = {}) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keys = key.split('.')

    return roll(keys, acc, value)
  }, type)
}

export const packChild = ({ name, obj }) => {
  const obj2 = {}
  let isArray
  Object.entries(obj).forEach(k => {
    if (name.length < k[0].length && k[0].indexOf(name) === 0) {
      const n = k[0].slice(name.length + 1)
      // must be true for all matching
      isArray = !isNaN(n.slice(0, n.indexOf('.')))
      obj2[n] = k[1]
    }
  })
  if (Object.keys(obj2).length) return pack(obj2, isArray ? [] : {})
  return null
}

export const removeChild = ({ name, obj = {} }) => {
  const obj2 = {}
  Object.entries(obj).forEach(k => {
    if (name.length >= k[0].length || k[0].indexOf(name) !== 0) {
      obj2[k[0]] = k[1]
    }
  })
  return obj2
}
