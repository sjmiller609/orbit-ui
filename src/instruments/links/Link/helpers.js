// check if url is external
export const externalUrl = url => {
  if (typeof url === 'object') return false
  const path = /^https?:\/\/|^\/\//i
  if (path.test(url) && !~url.indexOf(window.location.origin)) return true
  return
}
