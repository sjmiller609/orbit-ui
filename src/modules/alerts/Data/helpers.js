import { formErrors } from 'instruments'

const errors = [
  {
    key: 'permissions',
    name: 'properties.alert_emails',
    error: 'Contact your workspace admin to upgrade your permissions.',
  },
]

export const handleError = error => formErrors({ error, errors })

export const trimError = error => error.split(':')[1].trim()
