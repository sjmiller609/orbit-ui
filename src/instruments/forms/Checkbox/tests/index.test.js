import React from 'react'
import { mountWrap } from 'helpers/tests'
import Checkbox from '../index'

const props = {
  type: 'radio',
  name: 'testDeployType',
  id: '1234',
  onChange: jest.fn(),
  onBlur: jest.fn(),
  validate: jest.fn(),
  required: true,
  label: React.Fragment,
  title: 'Test Checkbox',
  error: React.Fragment,
  className: 'testClassString',
  updateErrors: jest.fn(),
  setRef: jest.fn(),
  value: 'Kubernetes',
  fieldId: 'testDeployType',
}

describe('Checkbox', () => {
  it('renders correctly', () => {
    const wrapper = mountWrap(<Checkbox {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
