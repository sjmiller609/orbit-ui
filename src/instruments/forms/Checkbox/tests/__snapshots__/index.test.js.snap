// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders correctly 1`] = `
<Field
  className="testClassString"
  error={Symbol(react.fragment)}
  fieldId="testDeployType"
  id="1234"
  label={Symbol(react.fragment)}
  name="testDeployType"
  onBlur={[Function]}
  onChange={[Function]}
  required={true}
  setRef={[Function]}
  title="Test Checkbox"
  type="radio"
  updateErrors={[Function]}
  validate={[Function]}
  value="Kubernetes"
>
  <Checkbox
    className="field required testClassString"
    error={null}
    fieldId="testDeployType"
    id="1234"
    label={
      <FormLabel
        id="1234"
        info={undefined}
      >
        Symbol(react.fragment)
      </FormLabel>
    }
    name="testDeployType"
    onBlur={null}
    onChange={[Function]}
    required={true}
    setRef={[Function]}
    setShowError={[Function]}
    showError={false}
    title="Test Checkbox"
    type="radio"
    updateErrors={[Function]}
    validate={[Function]}
    value="Kubernetes"
  >
    <Row
      align="flex-start"
      auto={true}
      className="checkbox field required testClassString"
      id="testDeployType"
      justify="flex-start"
    >
      <Flex
        align="flex-start"
        auto={true}
        className="checkbox field required testClassString"
        flow="row"
        id="testDeployType"
        justify="flex-start"
      >
        <div
          className="flex row flex-start align-flex-start auto checkbox field required testClassString"
          id="testDeployType"
        >
          <input
            checked={true}
            className="input"
            name="testDeployType"
            onBlur={null}
            onChange={[Function]}
            required={true}
            title="Test Checkbox"
            type="checkbox"
            value="Kubernetes"
          />
          <div
            className="box on"
            onClick={[Function]}
          >
            <div
              className="check"
            />
          </div>
          <FormLabel
            id="1234"
          >
            <label
              className="label"
              htmlFor="1234"
            >
              <Info />
            </label>
          </FormLabel>
        </div>
      </Flex>
    </Row>
  </Checkbox>
</Field>
`;
